# This patch combines:
# Idc4c1e722c8e553041fc32ddbecdcab524bea64a, Add-hostname-to-merger-path-for-git-repos
# I6cb61bc296eee336918c4c116b5405eed21f01e7, Fix 'Add hostname to merger path'
# Id89373f1e1818368062e57c4806cc1176a9f0172, Add hostname to ZUUL_URL

diff --git a/zuul/merger/merger.py b/zuul/merger/merger.py
index b3cfaca..c53b9a4 100644
--- a/zuul/merger/merger.py
+++ b/zuul/merger/merger.py
@@ -16,6 +16,7 @@
 import git
 import os
 import logging
+import urlparse
 
 import zuul.model
 
@@ -199,7 +200,8 @@ class Repo(object):
 class Merger(object):
     log = logging.getLogger("zuul.Merger")
 
-    def __init__(self, working_root, connections, email, username):
+    def __init__(self, working_root, connections, email, username,
+                 append_hostname):
         self.repos = {}
         self.working_root = working_root
         if not os.path.exists(working_root):
@@ -207,6 +209,7 @@ class Merger(object):
         self._makeSSHWrappers(working_root, connections)
         self.email = email
         self.username = username
+        self.append_hostname = append_hostname
 
     def _makeSSHWrappers(self, working_root, connections):
         for connection_name, connection in connections.items():
@@ -226,7 +229,17 @@ class Merger(object):
     def addProject(self, project, url):
         repo = None
         try:
-            path = os.path.join(self.working_root, project)
+            parsed = None
+            root = self.working_root
+
+            if self.append_hostname:
+                parsed = urlparse.urlparse(url)
+            if parsed:
+                root = os.path.join(self.working_root, parsed.hostname)
+            elif self.append_hostname:
+                self.log.warning("Unable to find hostname in url: %s" % url)
+
+            path = os.path.join(root, project)
             repo = Repo(url, path, self.email, self.username)
 
             self.repos[project] = repo
diff --git a/zuul/merger/server.py b/zuul/merger/server.py
index d56993c..f441003 100644
--- a/zuul/merger/server.py
+++ b/zuul/merger/server.py
@@ -44,8 +44,14 @@ class MergeServer(object):
         else:
             merge_name = None
 
+        if self.config.has_option('merger', 'append_hostname'):
+            append_hostname = self.config.getboolean(
+                'merger', 'append_hostname')
+        else:
+            append_hostname = False
+
         self.merger = merger.Merger(merge_root, connections, merge_email,
-                                    merge_name)
+                                    merge_name, append_hostname)
 
     def start(self):
         self._running = True
diff --git a/zuul/merger/merger.py b/zuul/merger/merger.py
index c53b9a4..d7dc2ad 100644
--- a/zuul/merger/merger.py
+++ b/zuul/merger/merger.py
@@ -242,14 +242,14 @@ class Merger(object):
             path = os.path.join(root, project)
             repo = Repo(url, path, self.email, self.username)
 
-            self.repos[project] = repo
+            self.repos["%s/%s" % (url, project)] = repo
         except Exception:
             self.log.exception("Unable to add project %s" % project)
         return repo
 
     def getRepo(self, project, url):
-        if project in self.repos:
-            return self.repos[project]
+        if "%s/%s" % (url, project) in self.repos:
+            return self.repos["%s/%s" % (url, project)]
         if not url:
             raise Exception("Unable to set up repo for project %s"
                             " without a url" % (project,))
@@ -343,7 +343,7 @@ class Merger(object):
         for key, mrc in recent.items():
             project, branch = key
             try:
-                repo = self.getRepo(project, None)
+                repo = self.getRepo(project, item['url'])
                 zuul_ref = branch + '/' + item['ref']
                 repo.createZuulRef(zuul_ref, mrc)
             except Exception:
diff --git a/zuul/launcher/gearman.py b/zuul/launcher/gearman.py
index 02f78fd..9f67822 100644
--- a/zuul/launcher/gearman.py
+++ b/zuul/launcher/gearman.py
@@ -287,6 +287,19 @@ class Gearman(object):
         params['ZUUL_PIPELINE'] = pipeline.name
         params['ZUUL_URL'] = item.current_build_set.zuul_url
         params['ZUUL_VOTING'] = job.voting and '1' or '0'
+
+        if self.config.has_option('merger', 'append_hostname') and \
+                self.config.getboolean('merger', 'append_hostname'):
+            if params['ZUUL_URL'][-1] == '/':
+                params['ZUUL_URL'] = params['ZUUL_URL'][:-1]
+            connection_hostname = self.config.get('connection %s' % (
+                item.pipeline.source.connection.connection_name),
+                'server')
+            self.log.debug('Appending hostname %s to ZUUL_URL' % (
+                connection_hostname))
+            params['ZUUL_URL'] = "%s/%s" % (params['ZUUL_URL'],
+                                            connection_hostname)
+
         if hasattr(item.change, 'refspec'):
             changes_str = '^'.join(
                 ['%s:%s:%s' % (i.change.project.name, i.change.branch,

-- 
2.7.3
